docs/cookbook/paulilang/pauli.md

```diff
@@ -0,0 +1,683 @@
+# Pauli Matrix Algebra
+
+In this example, we'll implement a simple dialect that allows you to write mathematical expressions using the [Pauli matrices](https://en.wikipedia.org/wiki/Pauli_matrices) $\sigma_x, \sigma_y$ and $\sigma_z$.
+We'll add some basic rewriting routines so that the IR is symbolically rewritten using the basic relations
+
+$$
+\sigma_i \sigma_j = \epsilon_{ijk} \sigma_k,
+$$
+
+and
+
+$$
+\sigma_i^2 = \mathbb{I_2}
+$$
+
+Given these relations, you can rewrite longer expressions at compile time, such that the resulting code that is executed becomes much simpler and the number of matrix operations that need to be performed is reduced.
+
+We'll start out with the basic definitions and then add more features to it as we go along.
+
+
+## Defining the dialect
+
+First off, we'll define the dialect statements.
+Each statement defining a Pauli operator will be returning a numpy array.
+We can keep things concise by defining the properties in one class and then inheriting from that class for each statement.
+
+Each statement that you want to use in your dialect needs to be registered to it.
+That's done by providing your dialect as an argument to the [`statement`][kirin.decl.statement].
+
+```python
+from numbers import Number
+import numpy as np
+
+from kirin import ir, lowering
+from kirin.types import PyClass
+from kirin.decl import statement, info
+
+
+dialect = ir.Dialect("pauli")  # (1)!
+
+
+@statement  # (2)!
+class PauliOperator(ir.Statement):
+    traits = frozenset({ir.Pure(), lowering.FromPythonCall()})  # (3)!
+    pre_factor: Number = info.attribute(default=1)  # (4)!
+    result: ir.ResultValue = info.result(PyClass(np.matrix))
+
+
+@statement(dialect=dialect)  # (5)!
+class X(PauliOperator):
+    pass
+
+@statement(dialect=dialect)
+class Y(PauliOperator):
+    pass
+
+@statement(dialect=dialect)
+class Z(PauliOperator):
+    pass
+
+
+@statement(dialect=dialect)
+class Id(PauliOperator):
+    pass
+```
+
+
+1. Define the actual dialect to which we want to register our statements.
+2. Notice how we are not registering the `PauliOperator`. That's because we don't actually want users to create an instant of it since it serves just as an abstract type. This is similar to how a user cannot create a number of abstract type `Number`, it's always a type such as `float` or `int`.
+3. All Pauli matrices will be pure since they can only return an object of one type. They are also created by a simple call.
```

Comment:
I feel this sentence is not very accurate here:

1. if we think the value of a Pauli operator is `np.matrix` then the output of `pauli.X()` is not always the same because of memory allocation behaviour and limited precision of X. So it is not pure.
2. if we think the value of a Pauli operator is **an unique** `np.matrix`  representing Pauli X operator (a.k.a singletons) then it is pure. Because `pauli.X()` always returns the same value.

Purity is about whether your output is deterministicly only depends on your input afterall, thus there is a subtle difference here. I believe your intend is the 2nd case.

Reply:
@Roger-luo I'd agree that we should have the second case here. However, since I'm pulling in a pre-factor into the definitions, we don't always get the same matrix, but rather the same matrix, but multiplied by a number.

So, technically, I'm not really implementing Pauli matrices here, but rather a product of one with a number, which means they're not singletons. Are they still pure then?

I know that's not a very clean implementation, but this way I can implement rewrites on binary multiplications only, which makes the example much easier.